### PR TITLE
New xx auth

### PR DESCRIPTION
**Pull request by LowEntropyBody**

### CODE REVIEW

File: packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+`
+)
+
+func jointVerify(data *DataForVerify) error {
+	// UDBPubkey
+	block, _ := pem.Decode([]byte(UDBCertEncoded))
+	cert, _ := x509.ParseCertificate(block.Bytes)
+	UDBPubkeyGo := cert.PublicKey.(*gorsa.PublicKey)
+	UDBPubkey := &rsa.PublicKey{PublicKey: *UDBPubkeyGo}
+
```
Comment (LowEntropyBody) [inline packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go:59]: Is the way to convert udb.crt to rsa's pubkey right?

Comment (LowEntropyBody) [reply packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go:59]: This this code for get Auth:

const (
	multiHashSize = 32
	multiHashSha  = multiHash.SHA2_256
)

func hashFile(file []byte) ([]byte, error) {
	h, err := multiHash.GetVariableHasher(multiHashSha, multiHashSize)
	if err != nil {
		return nil, err
	}

	h.Write(file)
	return h.Sum(nil), nil
}


func GetAuth() {
	// File hash
	const file = "AQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGRobHB0eHyAhIiMkJSYnKCkqKywtLi8wMTIzNDU2Nzg5Ojs8PT4/QEFCQ0RFRkdISUpLTE1OT1BRUlNUVVZXWFlaW1xdXl9gYWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXp7fH1+f4CBgoOEhYaHiImKi4yNjo+QkZKTlJWWl5iZmpucnZ6foKGio6SlpqeoqaqrrK2ur7CxsrO0tba3uLm6u7y9vr/AwcLDxMXGx8jJysvMzc7P0NHS09TV1tfY2drb3N3e3+Dh4uPk5ebn6Onq6+zt7u/w8fLz9PX29/j5+vv8/f4AAQIDBAUGBwg="
	fileTrue, _ := base64.StdEncoding.DecodeString(file)
	fileHashBytes, _ := hashFile(fileTrue)
	fileHash := base64.StdEncoding.EncodeToString(fileHashBytes)

	// user pubkey
	const userPubKeyString = "-----BEGIN RSA PUBLIC KEY-----\nMIGJAoGBAKF4XwtFOAJYw7fDeTWBezUBT/HrkGCzUNXrs90zwiTw1NXR5PAgQZc1\ntjGb/W3IzNu44+SEXuHluyPdXtNbBZr3nM9umB23VCnPa6+0vEB/1hY8IjMHKn53\nNAk4tGsUrbV4q0OOSB8ltz7fXvTUX9X8TNYrFEDasO0JYpIKswhHAgMBAAE=\n-----END RSA PUBLIC KEY-----"
	userPubKey := base64.StdEncoding.EncodeToString([]byte(userPubKeyString))

	// Other params
	const timestamp = "1660777707027403466"
	const uploadSig = "LIc3g4j2twvNZSpwjhue3AOkuavy62sYLwlZyOH2Bp8lBX8LLReZune9pCaaq4mitGPpm5LSdmGxGn0w5c+AJrKDrQc2d5jWMCli7ZBNbwUu2TIdJIUMxcq1c99PI2R6ArT8w2YXvsmf7vPA2Nq6Hk/jt8oTihYCqtl1dfvkNxewoywr6dyGoULbzhNBjsIJnBGNUxFDOfkrHa2uTo4dTBrkZbn8weR/CtkDQCOt2MmC38Jl6cmHCTgVpCtN556DtKkhiIGeChKvrh0ehLDuupNnzzqxNGtXF8MIWPDtT+o3G/qW7s2Qd6UTdbxR39EuBosCd71CVsSBeNMIpahwUR5M0NIhscKVUn0kISc1qHXAwA4+QsfsP0PqPGe3U4bIux2cVvu6HNE4XTYdnb9fd7/8Pnu8b5ZnuUN9qNVaAnyw1jFTni9b/OQy581kkbrdMZAROynP3CVSQ0LZbNsSnKZ5HsZ26iFsGbECLuAmx5zIP6kPs58XQMtm5mw5/UNrEKxO1LQEf21887W8BeghNE3fkPxm5CuErQEmOup4IljxdBcQbrHpOfE+FXJm8oXdf6mNlm4btafufktiIFwmxtXZFvNYsW+34B1ZuqMfSb5w3K7PrqMg49wjWqrRC3pJCtgj1hC5kobc5MfN+/gblz1qjb+u+1iJbIodwL7PdAE="
	const userName = "AQIDBAUGBwgJCgsMDQ4PEBESExQVFhcYGQ=="
	const verifSig = "GyO2ON3/FJECK/M1/qRoCH15HCbz1xGEI1qXvq9Zp/WZsk5/RtoSOJRA4E7oJoIOdBgjOG+xD+YPvZhPINda6gM0KJxhFr4Y8IUTko5mB0Fl/uGZZK/YWjUDYCrIlpUx994FMt97QhW7oQiftrhu1bTuMgOwLK7rSaFh0iz8Vp9j4aZSGlAtYANHc9kyBITd9GJOPZpFwDMLx9PtTGAiu9b8Fz2fq5KUANzCgtWqbfwwHCoO5+ue2yYNSYc3fedQ8uWP03WAhpodZ/0HKq/dFt238tbTTRYb5MXIs6CV0pwDevfP7F5O1CesZmG/7ySRvfYWeSi58T7UgMkJpI4xuLGo3NlqgGvp8nATX1/GEhYLdO8rPPk0cc4vdvw+aUqSZzuMcR/UioagdOUMYO59/ne/jiIe7kYlGiTe50bKjh8INiprKRTBad1OWRaRWj9YfE2yCqebvkQQ/+2lgzSYbYu1U0+93j+kChmCcoXH8RbCIazT9AEbMBNF5bvbubA52ql6o1tBF4fqBnpNGK6ad7Zn0tr68vmHrgL+2nR+L0lrW5VFUfpsGVOl4ULPO5czKK5yh7WNpuRjgQzAzj/0ZwT5rsZ4CmsFCxqaMeRB4YdZOgbcJmzqkWvsU7Yrd9OGISKvTQM+7i7ZjKY+RyyWbGA2VHwe9bDaDgQ7fDZuQ3M="

	auth := base64.StdEncoding.EncodeToString([]byte("xx-" + userPubKey + "-" + userName + "-" +
		fileHash + "-" + timestamp + "-" + uploadSig + ":" + verifSig))
	fmt.Println(auth)
}

File: packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	// Get timestamp
+	timestampInt64, err := strconv.ParseInt(data.Timestamp, 10, 64)
+	if err != nil {
+		return err
+	}
+	timestamp := time.Unix(0, timestampInt64)
+
+	if err := verifyUpload(userPubKey, time.Now(), timestamp, fileHash, uploadSig); err != nil {
```
Comment (LowEntropyBody) [inline packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go:104]: use userpubkey for validation?

File: packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	// Get verifSig
+	verifSig, err := base64.StdEncoding.DecodeString(data.VerifSig)
+	if err != nil {
+		return err
+	}
+
+	// Verify user pubkey and name
+	if err := verifyVerification(UDBPubkey, userNameHash, userPubKey, verifSig); err != nil {
```
Comment (LowEntropyBody) [inline packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go:81]: use UDB pubkey for validation?

File: packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go (hunk 1 above)
Comment (joshem) [reply packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go:59]: Yeah, we wrote a way to do it in our crypto library found here: gitlab.com/xx_network/crypto/tls
Here is an example of how to do so:

`	cert, err := tls.LoadCertificate(UDBCertEncoded)
	if err != nil {
		return err
	}
	
	udPubKey, err := tls.ExtractPublicKey(cert)
	if err != nil {
		return err

	}`

File: packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go (hunk 3 above)
Comment (joshem) [reply packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go:81]: Yeah, this signature is from our UD server

File: packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go (hunk 2 above)
Comment (joshem) [reply packages/ipfs-w3auth/auth-xx/validator/src/jointVerification.go:104]: Yeah, this signature comes from the clinet

**Review by BOBOKaa — APPROVED** [review]

**Pull request merged**